apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.CommonTesting/SpannerTestDatabaseBase.cs

```diff
@@ -0,0 +1,164 @@
+
+    public virtual string SpannerHost { get; } = GetEnvironmentVariableOrDefault("TEST_SPANNER_HOST", null);
+    public virtual string SpannerPort { get; } = GetEnvironmentVariableOrDefault("TEST_SPANNER_PORT", null);
+    public virtual string SpannerInstance { get; } = GetEnvironmentVariableOrDefault("TEST_SPANNER_INSTANCE", "spannerintegration");
+    public abstract string SpannerDatabase { get; }
+
+    /// <summary>
+    /// Returns true if the database was created just for this test, or false if the database was an existing one
+    /// specified through an environment variable.
+    /// </summary>
+    public abstract bool Fresh { get; }
+
+    // Connection string including database, generated from the above properties
+    public string ConnectionString { get; protected set; }
+    // Connection string without the database, generated from the above properties
+    public string NoDbConnectionString { get; protected set; }
+    public string ProjectId { get; protected set; }
+    public DatabaseName DatabaseName { get; protected set; }
+    internal SpannerClientCreationOptions SpannerClientCreationOptions { get; set; }
+
+    protected static string GetEnvironmentVariableOrDefault(string name, string defaultValue)
+    {
+        string value = Environment.GetEnvironmentVariable(name);
+        return string.IsNullOrEmpty(value) ? defaultValue : value;
+    }
+    
+    protected virtual void CreateInstance(string projectId)
+    {
+        // Try to create an instance and ignore any AlreadyExists error.
+        var instanceAdminClient = InstanceAdminClient.Create();
+        var instanceName = InstanceName.FromProjectInstance(projectId, SpannerInstance);
+        try
+        {
+            // Initialize request parameters.
+            Instance instance = new Instance
+            {
+                InstanceName = instanceName,
+                // TODO: Check if this instance id should come from environment variable or should have any other value. 
+                ConfigAsInstanceConfigName = InstanceConfigName.FromProjectInstanceConfig(projectId, "regional-us-central1"),
+                DisplayName = "Test Instance.",
+                NodeCount = 1
+            };
+            ProjectName projectName = ProjectName.FromProject(projectId);
+            var completedResponse = instanceAdminClient.CreateInstance(projectName, instanceName.InstanceId, instance).PollUntilCompleted();
+
+            if (completedResponse.IsFaulted)
+            {
+                Logger.DefaultLogger.Debug($"Error while creating instance: {completedResponse.Exception}");
+                throw completedResponse.Exception;
+            }
+
+            Logger.DefaultLogger.Debug($"Instance {SpannerInstance} created successfully.");
+        }
+        catch (RpcException e) when (e.StatusCode == StatusCode.AlreadyExists)
+        {
+            // Ignore
+        }
+    }
+
+    protected virtual void CreatePostgreSqlDatabase()
```

Comment:
Is there a reason this is in the base class? It feels odd to have PG-specific code in the base class.

Would it make sense to have a method here that accepts the DatabaseDialect as a parameter, to avoid duplication?